Added .cpp implementation of the Radix Sort Algorithm

adithya-menon-r opened the pull request:
The changes have been made. This PR solves #59 

Please review @Ashrockzzz2003

mdxaasil commented:
It seems that most of the code might be AI-generated. Could you please provide proof to confirm or disprove this? @adithya-menon-r

adithya-menon-r commented:
@mdxaasil I learnt the Radix Sort Algorithm from Jenny's Lectures. I saw both her Count Sort & Radix Sort videos to understand the algorithm. 

I then implemented it in .cpp using the logic I learnt, whenever I got stuck with the implementation I went back to the videos and also referred Programiz for help. 

For example, when modifying the bucket array with the cumulative frequency, I used a new variable called sum and then kept summing the elements and then assigning it back, but I found a way better and efficient implementation online which I never thought of -> `count[i] += count [i - 1]`. This is what I have included now. 

I have also implemented the algorithm using the standard int array found in C. Since I am still relatively new to C++, I chose not to use a vector, which is generally what I have seen LLMs use in .cpp code.

I also saw Ashwin anna's msg in the group about a PR with well formatted .cpp code. So, I took some time to try and do the same. 

If you would like me to explain the code, I would be happy to do so. 

Thank You

mdxaasil commented:
Great then. Just add a few more random test cases instead of only non-increasing one. Rest is good :)

adithya-menon-r commented:
@mdxaasil I have added 3 unsorted arrays with random numbers for the test cases. Please review it. Thanks!

mdxaasil commented:
@adithya-menon-r Can you use a random function instead of manually writing?

Ashrockzzz2003 commented on sorting/radix_sort/radix_sort.cpp:
```diff
@@ -0,0 +1,96 @@
+
+using namespace std;
+
+// Gets the maximum element in array (To find the max number of digits)
```
Comments can end with a `full-stop`. Will improve readability of code.
Here and everywhere else in code.
https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

Ashrockzzz2003 commented on sorting/radix_sort/radix_sort.cpp:
```diff
@@ -0,0 +1,96 @@
+using namespace std;
+
+// Gets the maximum element in array (To find the max number of digits)
+int getMax(int arr[], int num) {
```
what if array's empty. May happen!

adithya-menon-r replied:
Should I just check if the array is empty and then exit the program if it is or should I use exception handling?

Ashrockzzz2003 commented on sorting/radix_sort/radix_sort.cpp:
```diff
@@ -0,0 +1,96 @@
+}
+
+int main() {
+    const int num = 100; // Number of elements in the unsorted array
```
instead of comment, can we make the var name `self-explanatory`

Ashrockzzz2003 commented on sorting/radix_sort/radix_sort.cpp:
```diff
@@ -0,0 +1,96 @@
+
+// Count Sort Algorithm
+void countSort(int arr[], int num, int pos) {
+    int count[10] = {0}; // Sorting numbers here, hence base 10
```
what does `count` represent? Use a more `self-explanatory` name.

Ashrockzzz2003 reviewed (changes requested):
Are negative numbers handled?

adithya-menon-r commented:
> Are negative numbers handled?

@Ashrockzzz2003 I haven't handled negative numbers. I'll fix it and make the requested changes.

adithya-menon-r commented:
> Are negative numbers handled?

@Ashrockzzz2003 I was thinking of 2 approaches for it... 

1. Find the most negative number and then shift all numbers in the array with that number, then sort it and then remove the shift from the sorted array. So, if the array is `[8, 3, -1, -3, 2]`, I would shift the array to `[11, 6, 2, 0, 5]`, sort it to `[0, 2, 5, 6, 11]` and then remove the shift and get `[-3, -1, 2, 3, 8]`.
2. Split the array into 2 seperate arrays, one with all negative numbers and the other with all the positive numbers, sort them independently, (the negative numbers array will be sorted by taking absolute value) and them merge them back.

Which approach do you think would be better? Thanks!

adithya-menon-r commented:
Hey @Ashrockzzz2003, I have made the requested changes. I used the element shift approach I mentioned above to handle negative numbers in the sorting algorithm. 

![image](https://github.com/user-attachments/assets/b130cf74-c668-44f2-aacf-82aadf27fc62)

Please review it and let me know if I have to make any more changes. Thanks!

Ashrockzzz2003 commented:
> Hey @Ashrockzzz2003, I have made the requested changes. I used the element shift approach I mentioned above to handle negative numbers in the sorting algorithm.
> 
> ![image](https://private-user-images.githubusercontent.com/115925529/399321989-b130cf74-c668-44f2-aacf-82aadf27fc62.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1NzM4MDgsIm5iZiI6MTczNTU3MzUwOCwicGF0aCI6Ii8xMTU5MjU1MjkvMzk5MzIxOTg5LWIxMzBjZjc0LWM2NjgtNDRmMi1hYWNmLTgyYWFkZjI3ZmM2Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIzMFQxNTQ1MDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ODZhNDM1NzRlNDFkNjM2NjkyMTA1ZDVmOTUwZTdiZDczOGRjMDMyMjkxZDNkM2U4ZGMxNDY1NzM1YTY4YzljJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.-9vvnr3hbP8P5GMDMIMQX7KWzlduyIaXoGGpTE-xMLg)
> 
> Please review it and let me know if I have to make any more changes. Thanks!

Hey @adithya-menon-r 
Sorry for replying late

The problem with the first approach - what if the number exceeds INT range after adding. Even if one integer exceeds INT_MAX after shifting, it'll break the code!

Second approach does not have this problem.

adithya-menon-r commented:
@Ashrockzzz2003 Ah... that makes sense. I will make the changes and use the second approach. Thanks!

adithya-menon-r commented:
Hey @Ashrockzzz2003, I have implemented the second approach to handle negative numbers. It splits the array into two different arrays, one with all the negative numbers (absolute value) and the positive numbers. The Radix Sort algorithm then sorts them independently and then the sorted arrays are merged back. 

I added this logic in a seperate function `sortArray()` which calls `radixSort()` as this is technically not part of the standard Radix Sort Implementation.

Since the number of positive and negative numbers are not known, I felt this was the perfect oppurtunity for me to learn how to use vectors in C++. The sorting logic now uses vector instead of the standard C array, but I decided to continue using the standard C array until the splitting in `sortArray()`. 

![image](https://github.com/user-attachments/assets/077232a7-b5ee-4b07-b12c-40f9dcd07989)

Please review it and let me know if I have to make any more changes. Thanks!

Ashrockzzz2003 commented on sorting/radix_sort/radix_sort.cpp:
```diff
@@ -0,0 +1,130 @@
+
+using namespace std;
+
+// Gets the maximum element in array (To find the max number of digits).
+int getMax(vector<int>& arr) {
+    if (arr.empty()) { 
+        cerr << "The Array to Sort is Empty!" << endl; 
+        exit(1);
+    }
+    int max = arr[0];
+    for (int num : arr) {
+        if (num > max) {
+            max = num;
+        }
+    }
+    return max;
+}
```
Now that you're using `vector`, there's an inbuilt method `*max_element`. Use that instead of writing your own!

Ashrockzzz2003 commented on sorting/radix_sort/radix_sort.cpp:
```diff
@@ -0,0 +1,130 @@
+
+int main() {
+    const int size = 20;
+    int arr[size];
```
Use a vector instead. Why mix up. Creates unnecessary confusions! Design in code is important.

Ashrockzzz2003 commented on sorting/radix_sort/radix_sort.cpp:
```diff
@@ -0,0 +1,130 @@
+}
+
+// Helper function to print elements of array.
+void printArray(int arr[], int size) {
```
instead of passing in the entire array as object. When you use vector below, pass reference pointer here. memory saved!

Ashrockzzz2003 reviewed (changes requested):
Good job handling negative numbers.

adithya-menon-r commented:
Hey @Ashrockzzz2003, I have made the requested changes. I have used vector everywhere and I have also used the `max_element` method instead of the custom `getMax()` method.

Please review it. Thanks!

Ashrockzzz2003 reviewed (approved):
LGTM
Awesome code quality @adithya-menon-r

Ashrockzzz2003 commented:
Add yourself and @akshayks13 to `README.md`

Akshay KS - Python refer #22

Ashrockzzz2003 commented:
/bounty 40 @adithya-menon-r

adithya-menon-r commented:
Hey @Ashrockzzz2003, I have made the requested changes in the `README.md` and created a PR for the same.

adithya-menon-r commented:
@Ashrockzzz2003 @mdxaasil, thank you for reviewing my PR. Learned a lot, not just about Radix Sort but also some really useful C++ concepts and how to improve comment formatting and design choices in code. Appreciate the feedback and guidance!
